File: C#/C#.NET/modul_1_grundlagen/LiveCoding/L01BubbleSort/BubbleSort.cs
https://upload.wikimedia.org/wikipedia/commons/c/c8/Bubble-sort-example-300px.gif

uint[] zahlen = [28, 26, 6, 4, 2];

uint platzhalter;

// Schritt 3: Wiederhole 2. solange bis alle Zahlen sortiert sind.
// Wir lassen also jede Zahl nach rechts "aufsteigen", bis diese an der richtigen Position ist.
// FRAGE: Wie oft müssen wir die innere und außere Schleife wiederholen um alle Zahlen sortiert zu haben, wenn wir 5 Zahlen zu sortiern haben?
for (int j = 0; j < zahlen.Length - 1; j++)
{
    //Console.WriteLine((j + 1) + " Durchlauf"); // string concatenation
    Console.WriteLine($"{j + 1} Durchlauf");   // string interpolation
    //Console.WriteLine("{0} Durchlauf", j + 1);   // formated output

    // Schritt 2: Wiederhole 1. für alle Paare mit Index 0 und 1, 1 und 2, 2 und 3, 3 und 4.
    for (int i = 0; i < zahlen.Length - 1 - j; i++)
    {
        // Schritt 1: Vergleiche die 1. (Index 0) und 2. (Index 1) Zahl im Array.
        // Falls die 1. größer ist wie die 2., dann vertausche diese, andernfalls mach nichts.
        if (zahlen[i] > zahlen[i + 1])
        {
            platzhalter = zahlen[i];
            zahlen[i] = zahlen[i + 1];
            zahlen[i + 1] = platzhalter;
        }

        Console.WriteLine($"{i + 1} Paar [{string.Join(",", zahlen)}]");
    }

    Console.WriteLine();
}

File: Python/bubblesort.py
zahlen = [6, 5, 3, 1, 8, 7, 2, 4]
anzahl_paare = len(zahlen) - 1

# 4. Lass mehrere Bubbles aufsteigen, bis wir sortiert sind.
for i in range(len(zahlen)):
    is_sorted = True

    # 3. Schiebe roten Slider (die Bubble) solange nach rechts bis wir am Ende sind.
    # for bubble in [0,1,2,3,4,5,6]: # range erzeugt uns [0,1,2,3,4,5,6]
    for bubble in range(anzahl_paare - i):
        # 2. vertausche wenn erste zahl größer zweite zahl
        if zahlen[bubble] > zahlen[bubble + 1]:
            # 1. vertausche zwei zahlen aus einer liste
            lostandfound = zahlen[bubble]
            zahlen[bubble] = zahlen[bubble + 1]
            zahlen[bubble + 1] = lostandfound
            is_sorted = False

            # 0. Greife auf Elmente der Liste zu
            # print(zahlen[0])
            # print(zahlen[1])
            # print(zahlen[0] > zahlen[1])

            # lostandfound = zahlen[0]
            # zahlen[0] = zahlen[1]
            # zahlen[1] = lostandfound
            # is_sorted = False

        print(zahlen)

    if is_sorted:
        break
    
    print()


    # Mit while

anzahl_paare = len(zahlen) - 1
is_sorted = True
repeated_bubbles = 0

# 4. Lass mehrere Bubbles aufsteigen, bis wir sortiert sind.
while is_sorted:
    # 3. Schiebe roten Slider (die Bubble) solange nach rechts bis wir am Ende sind.
    for bubble in range(anzahl_paare - repeated_bubbles):
        # 2. vertausche wenn erste zahl größer zweite zahl
        if zahlen[bubble] > zahlen[bubble + 1]:
            # 1. vertausche zwei zahlen aus einer liste
            lostandfound = zahlen[bubble]
            zahlen[bubble] = zahlen[bubble + 1]
            zahlen[bubble + 1] = lostandfound
            is_sorted = False

        print(zahlen)    
    print()

    repeated_bubbles += 1
